ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,625 @@
+        if (peer >= sg_size) {
+            continue;
+        }
+        if (peer >= end_peer) {
+            peer = (((peer + cur_base - sg_start) % sg_size) + sg_start);
+        } else {
+            peer = peer + cur_base;
+        }
+        /* Checks to ensure that the sends are limited to the necessary ones.
+           It also ensures 'preq' not exceeding the max allocated. */
+        if ((peer < size) && (peer != rank) && (peer != cur_base)) {
+            *nreqs = *nreqs + 1;
+            err = MCA_PML_CALL(isend(buff, count, datatype, peer, MCA_COLL_BASE_TAG_ALLGATHER,
+                                     MCA_PML_BASE_SEND_STANDARD, comm, preq++));
+            if (MPI_SUCCESS != err) {
+                return err;
+            }
+        }
+    }
+
+    return err;
+}
+
+static inline int lin_sg_bcast_intra(void *buff, size_t count, struct ompi_datatype_t *datatype,
+                                     int rank, int dim, int size, int sg_size, int cur_base,
+                                     int sg_start, struct ompi_communicator_t *comm,
+                                     mca_coll_base_module_t *module, ompi_request_t **preq,
+                                     int *nreqs)
+{
+    int peer;
+    int err;
+    int sg_end;
+
+    sg_end = sg_start + sg_size - 1;
+    if (sg_end >= size) {
+        sg_end = size - 1;
+    }
+
+    if (rank == cur_base) {
+        for (peer = sg_start; peer <= sg_end; peer++) {
+            if (peer == cur_base) {
+                continue;
+            }
+            *nreqs = *nreqs + 1;
+            err = MCA_PML_CALL(isend(buff, count, datatype, peer, MCA_COLL_BASE_TAG_ALLGATHER,
+                                     MCA_PML_BASE_SEND_STANDARD, comm, preq++));
+            if (MPI_SUCCESS != err) {
+                return err;
+            }
+        }
+    } else {
+        err = MCA_PML_CALL(recv(buff, count, datatype, cur_base, MCA_COLL_BASE_TAG_ALLGATHER, comm,
+                                MPI_STATUS_IGNORE));
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+    }
+
+    return err;
+}
+
+/*
+ * sg_bcast_intra
+ *
+ * Function:    broadcast operation within a subgroup
+ * Accepts:     Arguments of MPI_Bcast() plus subgroup params
+ * Returns:     MPI_SUCCESS or error code
+ *
+ * Description: O(N) or O(log(N)) algorithm based on count.
+ *
+ * Memory:      No additional memory requirements beyond user-supplied buffers.
+ *
+ */
+static inline int sg_bcast_intra(void *buff, size_t count, struct ompi_datatype_t *datatype, int rank,
+                                 int dim, int size, int sg_size, int cur_base, int sg_start,
+                                 struct ompi_communicator_t *comm, mca_coll_base_module_t *module,
+                                 ompi_request_t **preq, int *nreqs)
+{
+    int err;
+    size_t total_dsize, dsize;
+
+    ompi_datatype_type_size(datatype, &dsize);
+    total_dsize = dsize * count;
+
+    if (total_dsize <= 8192) {
```

Comment:
It seems like `8192` comes up often here. Why was that chosen? Is that related to cache sizes? Should this be a `#define` constant?

Reply:
These numbers / conditions are empirically derived. I don't think we should #define these.